### PR TITLE
desktop: Implement crash report dialog when a panic happens

### PR DESCRIPTION
**Pull request by Dinnerbone**
When a panic happens:

image

Pressing Yes takes you to [this new issue template, prefilled](https://github.com/dinnerbone/ruffle/issues/new?assignees=&labels=bug&template=crash_report.yml&panic_text=panicked+at+%27wgpu+error%3A+Validation+Error%0A%0ACaused+by%3A%0A++++In+Device%3A%3Acreate_texture%0A++++Dimension+X+value+390000+exceeds+the+limit+of+16384%0A%0A%27%2C+C%3A%5CUsers%5CDinnerbone%5C.cargo%5Cgit%5Ccheckouts%5Cwgpu-e4487c3622394142%5Ca154700%5Cwgpu%5Csrc%5Cbackend%5Cdirect.rs%3A2962%3A5&operating_system=Windows+10.0.19044+%28Windows+10+Pro%29+%5B64-bit%5D&extra_info=Renderer%3A+wgpu%0AAdapter+Backend%3A+Vulkan%0AAdapter+Name%3A+%22AMD+Radeon+RX+6900+XT%22%0AAdapter+Device+Type%3A+DiscreteGpu%0AAdapter+Driver+Name%3A+%22AMD+proprietary+driver%22%0AAdapter+Driver+Info%3A+%2222.11.2%22%0AEnabled+features%3A+PUSH_CONSTANTS%0AAvailable+features%3A+DEPTH_CLIP_CONTROL+%7C+DEPTH32FLOAT_STENCIL8+%7C+TEXTURE_COMPRESSION_BC+%7C+INDIRECT_FIRST_INSTANCE+%7C+TIMESTAMP_QUERY+%7C+PIPELINE_STATISTICS_QUERY+%7C+SHADER_FLOAT16+%7C+MAPPABLE_PRIMARY_BUFFERS+%7C+TEXTURE_BINDING_ARRAY+%7C+BUFFER_BINDING_ARRAY+%7C+STORAGE_RESOURCE_BINDING_ARRAY+%7C+SAMPLED_TEXTURE_AND_STORAGE_BUFFER_ARRAY_NON_UNIFORM_INDEXING+%7C+UNIFORM_BUFFER_AND_STORAGE_TEXTURE_ARRAY_NON_UNIFORM_INDEXING+%7C+PARTIALLY_BOUND_BINDING_ARRAY+%7C+MULTI_DRAW_INDIRECT+%7C+MULTI_DRAW_INDIRECT_COUNT+%7C+ADDRESS_MODE_CLAMP_TO_BORDER+%7C+POLYGON_MODE_LINE+%7C+POLYGON_MODE_POINT+%7C+TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES+%7C+SHADER_FLOAT64+%7C+CONSERVATIVE_RASTERIZATION+%7C+VERTEX_WRITABLE_STORAGE+%7C+CLEAR_TEXTURE+%7C+SPIRV_SHADER_PASSTHROUGH+%7C+SHADER_PRIMITIVE_INDEX+%7C+MULTIVIEW+%7C+TEXTURE_FORMAT_16BIT_NORM+%7C+ADDRESS_MODE_CLAMP_TO_ZERO+%7C+WRITE_TIMESTAMP_INSIDE_PASSES%0ACurrent+limits%3A+Limits+%7B+max_texture_dimension_1d%3A+16384%2C+max_texture_dimension_2d%3A+16384%2C+max_texture_dimension_3d%3A+8192%2C+max_texture_array_layers%3A+256%2C+max_bind_groups%3A+4%2C+max_bindings_per_bind_group%3A+640%2C+max_dynamic_uniform_buffers_per_pipeline_layout%3A+8%2C+max_dynamic_storage_buffers_per_pipeline_layout%3A+0%2C+max_sampled_textures_per_shader_stage%3A+16%2C+max_samplers_per_shader_stage%3A+16%2C+max_storage_buffers_per_shader_stage%3A+4294967295%2C+max_storage_textures_per_shader_stage%3A+0%2C+max_uniform_buffers_per_shader_stage%3A+11%2C+max_uniform_buffer_binding_size%3A+16384%2C+max_storage_buffer_binding_size%3A+2147483648%2C+max_vertex_buffers%3A+8%2C+max_vertex_attributes%3A+16%2C+max_vertex_buffer_array_stride%3A+255%2C+max_push_constant_size%3A+96%2C+min_uniform_buffer_offset_alignment%3A+32%2C+min_storage_buffer_offset_alignment%3A+32%2C+max_inter_stage_shader_components%3A+60%2C+max_compute_workgroup_storage_size%3A+0%2C+max_compute_invocations_per_workgroup%3A+0%2C+max_compute_workgroup_size_x%3A+0%2C+max_compute_workgroup_size_y%3A+0%2C+max_compute_workgroup_size_z%3A+0%2C+max_compute_workgroups_per_dimension%3A+0%2C+max_buffer_size%3A+268435456+%7D%0ASurface+samples%3A+4%0ASurface+size%3A+Extent3d+%7B+width%3A+800%2C+height%3A+600%2C+depth_or_array_layers%3A+1+%7D%0ASWF%3A%20catburglar.swf)

### CODE REVIEW

File: desktop/src/main.rs [hunk 1]
```diff
@@ -868,12 +874,58 @@ fn init() {
...
+        });
+        SWF_INFO.with(|i| {
+            if let Some(swf_name) = i.take() {
+                extra_info.push(format!("SWF: {swf_name}"));
+            }
+        });
+        if !extra_info.is_empty() {
+            params.push(("extra_info", extra_info.join("\n")));
+        }
+        if let Ok(url) = Url::parse_with_params("https://github.com/dinnerbone/ruffle/issues/new?assignees=&labels=bug&template=crash_report.yml", &params) {
```
Comment (n0samu) [inline desktop/src/main.rs:925]: This should be ruffle-rs/ruffle 😄

Comment (Dinnerbone) [reply desktop/src/main.rs:925]: Haha yes it should! I changed it right before committing to get that link <_<

**Comment by Bale001** [comment]
We should also include the AVM2 stack trace in the crash report, which will be helpful for AS3 bugs

**Comment by Bale001** [comment]
> @Bale001 how safe is our callback collection? Is there any possibility of it panicking during access?

It's safe to access - `StaticCallstack::avm2` will simply do nothing if it can't access the callstack

**Comment by Dinnerbone** [comment]
[New example form](https://github.com/dinnerbone/ruffle/issues/new?assignees=&labels=bug&template=crash_report.yml&panic_text=panicked+at+%27attempt+to+divide+by+zero%27%2C+core%5Csrc%5Cavm2%5Cglobals%5Cmath.rs%3A75%3A13&platform=Desktop+app&operating_system=Windows+10.0.19044+%28Windows+10+Pro%29+%5B64-bit%5D&ruffle_version=0.1.0-nightly+%28b5c6d5deb+2023-01-17%29&extra_info=Renderer%3A+wgpu%0AAdapter+Backend%3A+Vulkan%0AAdapter+Name%3A+%22AMD+Radeon+RX+6900+XT%22%0AAdapter+Device+Type%3A+DiscreteGpu%0AAdapter+Driver+Name%3A+%22AMD+proprietary+driver%22%0AAdapter+Driver+Info%3A+%2222.11.2%22%0AEnabled+features%3A+PUSH_CONSTANTS%0AAvailable+features%3A+DEPTH_CLIP_CONTROL+%7C+DEPTH32FLOAT_STENCIL8+%7C+TEXTURE_COMPRESSION_BC+%7C+INDIRECT_FIRST_INSTANCE+%7C+TIMESTAMP_QUERY+%7C+PIPELINE_STATISTICS_QUERY+%7C+SHADER_FLOAT16+%7C+MAPPABLE_PRIMARY_BUFFERS+%7C+TEXTURE_BINDING_ARRAY+%7C+BUFFER_BINDING_ARRAY+%7C+STORAGE_RESOURCE_BINDING_ARRAY+%7C+SAMPLED_TEXTURE_AND_STORAGE_BUFFER_ARRAY_NON_UNIFORM_INDEXING+%7C+UNIFORM_BUFFER_AND_STORAGE_TEXTURE_ARRAY_NON_UNIFORM_INDEXING+%7C+PARTIALLY_BOUND_BINDING_ARRAY+%7C+MULTI_DRAW_INDIRECT+%7C+MULTI_DRAW_INDIRECT_COUNT+%7C+ADDRESS_MODE_CLAMP_TO_BORDER+%7C+POLYGON_MODE_LINE+%7C+POLYGON_MODE_POINT+%7C+TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES+%7C+SHADER_FLOAT64+%7C+CONSERVATIVE_RASTERIZATION+%7C+VERTEX_WRITABLE_STORAGE+%7C+CLEAR_TEXTURE+%7C+SPIRV_SHADER_PASSTHROUGH+%7C+SHADER_PRIMITIVE_INDEX+%7C+MULTIVIEW+%7C+TEXTURE_FORMAT_16BIT_NORM+%7C+ADDRESS_MODE_CLAMP_TO_ZERO+%7C+WRITE_TIMESTAMP_INSIDE_PASSES%0ACurrent+limits%3A+Limits+%7B+max_texture_dimension_1d%3A+16384%2C+max_texture_dimension_2d%3A+16384%2C+max_texture_dimension_3d%3A+8192%2C+max_texture_array_layers%3A+256%2C+max_bind_groups%3A+4%2C+max_bindings_per_bind_group%3A+640%2C+max_dynamic_uniform_buffers_per_pipeline_layout%3A+8%2C+max_dynamic_storage_buffers_per_pipeline_layout%3A+0%2C+max_sampled_textures_per_shader_stage%3A+16%2C+max_samplers_per_shader_stage%3A+16%2C+max_storage_buffers_per_shader_stage%3A+4294967295%2C+max_storage_textures_per_shader_stage%3A+0%2C+max_uniform_buffers_per_shader_stage%3A+11%2C+max_uniform_buffer_binding_size%3A+16384%2C+max_storage_buffer_binding_size%3A+2147483648%2C+max_vertex_buffers%3A+8%2C+max_vertex_attributes%3A+16%2C+max_vertex_buffer_array_stride%3A+255%2C+max_push_constant_size%3A+96%2C+min_uniform_buffer_offset_alignment%3A+32%2C+min_storage_buffer_offset_alignment%3A+32%2C+max_inter_stage_shader_components%3A+60%2C+max_compute_workgroup_storage_size%3A+0%2C+max_compute_invocations_per_workgroup%3A+0%2C+max_compute_workgroup_size_x%3A+0%2C+max_compute_workgroup_size_y%3A+0%2C+max_compute_workgroup_size_z%3A+0%2C+max_compute_workgroups_per_dimension%3A+0%2C+max_buffer_size%3A+268435456+%7D%0ASurface+samples%3A+4%0ASurface+size%3A+Extent3d+%7B+width%3A+800%2C+height%3A+600%2C+depth_or_array_layers%3A+1+%7D%0ASWF%3A+SQyvNBxOG2f1Uq.swf%0AAVM2+Callstack%3A%0A%60%60%60%0A%09at+Math%2FMath%3A%3Amax%28%29%0A%09at+flipline.api.assets%3A%3AAdSystem%24%2FprogressBarUpdate%28%29%0A%60%60%60)

**Comment by firoball** [comment]
Small thing: Is it really wanted that "additional information" is already pre-filled with an example in the form? People might overlook (or not even care) and just submit whatever is already listed there (Radeon adapter etc.). When reviewing the issue later on, this could cause confusions.

**Comment by Dinnerbone** [comment]
They can't just hit submit as some fields are required but empty. If we don't prefill system info, I'm not sure we'd ever get that info to honest

**Comment by firoball** [comment]
> They can't just hit submit as some fields are required but empty. If we don't prefill system info, I'm not sure we'd ever get that info to honest

Point taken. Maybe add "Example:" (or similar) in the default contents, to at least have a chance identifying lazy reports containing default text?

**Comment by Dinnerbone** [comment]
> Small thing: Is it really wanted that "additional information" is already pre-filled with an example in the form? People might overlook (or not even care) and just submit whatever is already listed there (Radeon adapter etc.). When reviewing the issue later on, this could cause confusions.

I just re-read this and maybe there's some confusion - the info prefilled *is the users info*, not example info. The link I gave as an example for this PR is what I got when I hit the "yes" button, but other users will get their own info.

**Comment by firoball** [comment]
> > Small thing: Is it really wanted that "additional information" is already pre-filled with an example in the form? People might overlook (or not even care) and just submit whatever is already listed there (Radeon adapter etc.). When reviewing the issue later on, this could cause confusions.
> 
> I just re-read this and maybe there's some confusion - the info prefilled _is the users info_, not example info. The link I gave as an example for this PR is what I got when I hit the "yes" button, but other users will get their own info.

Ah, got it. So I am seeing your system details only since it is preview for this PR.
No issue then. Sorry for the confusion.

**Review by Bale001 — APPROVED** [review]

**Comment by Dinnerbone** [comment]
[New example](https://github.com/dinnerbone/ruffle/issues/new?assignees=&labels=bug&template=crash_report.yml&panic_text=panicked+at+%27attempt+to+divide+by+zero%27%2C+core%5Csrc%5Cavm2%5Cglobals%5Cmath.rs%3A75%3A13&platform=Desktop+app&operating_system=Windows+10.0.19044+%28Windows+10+Pro%29+%5B64-bit%5D&ruffle_version=0.1.0-nightly+%28b5c6d5deb+2023-01-17%29&title=Crash+on+SQyvNBxOG2f1Uq.swf&extra_info=%23+AVM2+Callstack%0A%60%60%60%0A%09at+Math%2FMath%3A%3Amax%28%29%0A%09at+flipline.api.assets%3A%3AAdSystem%24%2FprogressBarUpdate%28%29%0A%60%60%60%0A%0A%23+Render+Info%0ARenderer%3A+wgpu%0AAdapter+Backend%3A+Vulkan%0AAdapter+Name%3A+%22AMD+Radeon+RX+6900+XT%22%0AAdapter+Device+Type%3A+DiscreteGpu%0AAdapter+Driver+Name%3A+%22AMD+proprietary+driver%22%0AAdapter+Driver+Info%3A+%2222.11.2%22%0AEnabled+features%3A+PUSH_CONSTANTS%0AAvailable+features%3A+DEPTH_CLIP_CONTROL+%7C+DEPTH32FLOAT_STENCIL8+%7C+TEXTURE_COMPRESSION_BC+%7C+INDIRECT_FIRST_INSTANCE+%7C+TIMESTAMP_QUERY+%7C+PIPELINE_STATISTICS_QUERY+%7C+SHADER_FLOAT16+%7C+MAPPABLE_PRIMARY_BUFFERS+%7C+TEXTURE_BINDING_ARRAY+%7C+BUFFER_BINDING_ARRAY+%7C+STORAGE_RESOURCE_BINDING_ARRAY+%7C+SAMPLED_TEXTURE_AND_STORAGE_BUFFER_ARRAY_NON_UNIFORM_INDEXING+%7C+UNIFORM_BUFFER_AND_STORAGE_TEXTURE_ARRAY_NON_UNIFORM_INDEXING+%7C+PARTIALLY_BOUND_BINDING_ARRAY+%7C+MULTI_DRAW_INDIRECT+%7C+MULTI_DRAW_INDIRECT_COUNT+%7C+ADDRESS_MODE_CLAMP_TO_BORDER+%7C+POLYGON_MODE_LINE+%7C+POLYGON_MODE_POINT+%7C+TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES+%7C+SHADER_FLOAT64+%7C+CONSERVATIVE_RASTERIZATION+%7C+VERTEX_WRITABLE_STORAGE+%7C+CLEAR_TEXTURE+%7C+SPIRV_SHADER_PASSTHROUGH+%7C+SHADER_PRIMITIVE_INDEX+%7C+MULTIVIEW+%7C+TEXTURE_FORMAT_16BIT_NORM+%7C+ADDRESS_MODE_CLAMP_TO_ZERO+%7C+WRITE_TIMESTAMP_INSIDE_PASSES%0ACurrent+limits%3A+Limits+%7B+max_texture_dimension_1d%3A+16384%2C+max_texture_dimension_2d%3A+16384%2C+max_texture_dimension_3d%3A+8192%2C+max_texture_array_layers%3A+256%2C+max_bind_groups%3A+4%2C+max_bindings_per_bind_group%3A+640%2C+max_dynamic_uniform_buffers_per_pipeline_layout%3A+8%2C+max_dynamic_storage_buffers_per_pipeline_layout%3A+0%2C+max_sampled_textures_per_shader_stage%3A+16%2C+max_samplers_per_shader_stage%3A+16%2C+max_storage_buffers_per_shader_stage%3A+4294967295%2C+max_storage_textures_per_shader_stage%3A+0%2C+max_uniform_buffers_per_shader_stage%3A+11%2C+max_uniform_buffer_binding_size%3A+16384%2C+max_storage_buffer_binding_size%3A+2147483648%2C+max_vertex_buffers%3A+8%2C+max_vertex_attributes%3A+16%2C+max_vertex_buffer_array_stride%3A+255%2C+max_push_constant_size%3A+96%2C+min_uniform_buffer_offset_alignment%3A+32%2C+min_storage_buffer_offset_alignment%3A+32%2C+max_inter_stage_shader_components%3A+60%2C+max_compute_workgroup_storage_size%3A+0%2C+max_compute_invocations_per_workgroup%3A+0%2C+max_compute_workgroup_size_x%3A+0%2C+max_compute_workgroup_size_y%3A+0%2C+max_compute_workgroup_size_z%3A+0%2C+max_compute_workgroups_per_dimension%3A+0%2C+max_buffer_size%3A+268435456+%7D%0ASurface+samples%3A+4%0ASurface+size%3A+Extent3d+%7B+width%3A+800%2C+height%3A+600%2C+depth_or_array_layers%3A+1+%7D%0A%0A%23+SWF%0AFilename%3A+SQyvNBxOG2f1Uq.swf%0A)

**Pull request merged**